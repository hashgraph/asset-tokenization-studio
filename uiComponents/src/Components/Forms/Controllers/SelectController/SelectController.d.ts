/// <reference types="react" />
import type { SelectProps } from "../../Select";
import type { Control, FieldValues, UseControllerProps } from "react-hook-form";
export interface SelectControllerProps<T extends FieldValues> extends Omit<SelectProps, "onChange" | "onBlur"> {
    control: Control<T>;
    id: UseControllerProps<T>["name"];
    rules?: Record<string, unknown>;
    showErrors?: boolean;
    onChange?: (value: SelectOption | SelectOption["value"]) => void;
    setsFullOption?: boolean;
}
export interface SelectOption<T = string, Y = string> {
    value: T;
    label: Y;
}
export declare const SelectController: <T extends FieldValues>(props: SelectControllerProps<T>) => JSX.Element;
