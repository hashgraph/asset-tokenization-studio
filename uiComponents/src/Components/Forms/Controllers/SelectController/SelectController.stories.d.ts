declare const meta: import("@storybook/types").ComponentAnnotations<import("@storybook/react/dist/types-0a347bb9").R, import("./SelectController").SelectControllerProps<import("react-hook-form").FieldValues>>;
export default meta;
export declare const NoValidations: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0a347bb9").R, import("./SelectController").SelectControllerProps<import("react-hook-form").FieldValues>>;
export declare const WithValidations: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0a347bb9").R, import("./SelectController").SelectControllerProps<import("react-hook-form").FieldValues>>;
export declare const HideErrors: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0a347bb9").R, import("./SelectController").SelectControllerProps<import("react-hook-form").FieldValues>>;
export declare const SetsFullOption: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0a347bb9").R, import("./SelectController").SelectControllerProps<import("react-hook-form").FieldValues>>;
