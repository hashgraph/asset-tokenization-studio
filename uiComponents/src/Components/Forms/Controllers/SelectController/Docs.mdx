import {
  Story,
  Meta,
  Title,
  ArgTypes,
  Primary,
  Stories,
  Source,
} from "@storybook/addon-docs";
import { Grid, GridItem, Text, Link, Center, Flex } from "@chakra-ui/react";
import * as SelectControllerStories from "./SelectController.stories";

<Meta of={SelectControllerStories} />

<Title />

<Primary />

## How to set up

SelectController is built with FieldController and Select (Input + Dropdown)

## Props

<ArgTypes of={SelectControllerStories} />

## Stories

<Stories title="" />
