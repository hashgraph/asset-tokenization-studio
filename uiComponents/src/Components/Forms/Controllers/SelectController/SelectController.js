var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { Select } from "../../Select";
import { omit as _omit } from "lodash";
import React, { useEffect, useState } from "react";
import { useController } from "react-hook-form";
import { FieldController } from "../FieldController/FieldController";
import _isEqual from "lodash/isEqual";
export var SelectController = function (props) {
    var _a;
    var control = props.control, id = props.id, options = props.options, rules = props.rules, _b = props.showErrors, showErrors = _b === void 0 ? true : _b, variant = props.variant, onChange = props.onChange, _c = props.setsFullOption, setsFullOption = _c === void 0 ? false : _c;
    var _d = useController({ name: id, control: control, rules: rules }), fieldState = _d.fieldState, _e = _d.field, onChangeDefault = _e.onChange, value = _e.value;
    var selectProps = _omit(props, ["control", "rules", "showErrors"]);
    var _f = useState(), selectedValue = _f[0], setSelectedValue = _f[1];
    // @ts-ignore
    var isRequired = !!(rules === null || rules === void 0 ? void 0 : rules["required"]) || !!((_a = rules === null || rules === void 0 ? void 0 : rules.validate) === null || _a === void 0 ? void 0 : _a["required"]);
    useEffect(function () {
        var newValue = options === null || options === void 0 ? void 0 : options.find(function (option) {
            if (value) {
                var valueToCompare = setsFullOption ? value.value : value;
                return _isEqual(option.value, valueToCompare);
            }
            return false;
        });
        setSelectedValue(newValue);
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [value, options]);
    return (React.createElement(FieldController, { fieldState: fieldState, errorMessageVariant: variant, showErrors: showErrors },
        React.createElement(Select, __assign({}, selectProps, { "data-testid": id, inputId: id, name: id, isRequired: isRequired, 
            // @ts-ignore
            onChange: function (val) {
                onChange === null || onChange === void 0 ? void 0 : onChange(setsFullOption ? val : val.value);
                onChangeDefault(setsFullOption ? val : val.value);
            }, value: selectedValue || "" }))));
};
