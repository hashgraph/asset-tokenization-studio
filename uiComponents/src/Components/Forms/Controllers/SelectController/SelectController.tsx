import type { SelectProps } from "../../Select";
import { Select } from "../../Select";
import { omit as _omit } from "lodash";
import React, { useEffect, useState } from "react";
import type { Control, FieldValues, UseControllerProps } from "react-hook-form";
import { useController } from "react-hook-form";
import { FieldController } from "../FieldController/FieldController";
import _isEqual from "lodash/isEqual";
export interface SelectControllerProps<T extends FieldValues>
  extends Omit<SelectProps, "onChange" | "onBlur"> {
  control: Control<T>;
  id: UseControllerProps<T>["name"];
  rules?: Record<string, unknown>;
  showErrors?: boolean;
  onChange?: (value: SelectOption | SelectOption["value"]) => void;
  setsFullOption?: boolean;
}

export interface SelectOption<T = string, Y = string> {
  value: T;
  label: Y;
}

export const SelectController = <T extends FieldValues>(
  props: SelectControllerProps<T>
) => {
  const {
    control,
    id,
    options,
    rules,
    showErrors = true,
    variant,
    onChange,
    setsFullOption = false,
  } = props;
  const {
    fieldState,
    field: { onChange: onChangeDefault, value },
  } = useController({ name: id, control, rules });
  const selectProps = _omit(props, ["control", "rules", "showErrors"]);
  const [selectedValue, setSelectedValue] = useState<SelectOption>();
  // @ts-ignore
  const isRequired = !!rules?.["required"] || !!rules?.validate?.["required"];

  useEffect(() => {
    const newValue = (options as SelectOption[])?.find(
      (option: SelectOption) => {
        if (value) {
          const valueToCompare = setsFullOption ? value.value : value;
          return _isEqual(option.value, valueToCompare);
        }
        return false;
      }
    );

    setSelectedValue(newValue);

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [value, options]);

  return (
    <FieldController
      fieldState={fieldState}
      errorMessageVariant={variant}
      showErrors={showErrors}
    >
      <Select
        {...selectProps}
        data-testid={id}
        inputId={id}
        name={id}
        isRequired={isRequired}
        // @ts-ignore
        onChange={(val: SelectOption) => {
          onChange?.(setsFullOption ? val : val.value);
          onChangeDefault(setsFullOption ? val : val.value);
        }}
        value={selectedValue || ""}
      />
    </FieldController>
  );
};
