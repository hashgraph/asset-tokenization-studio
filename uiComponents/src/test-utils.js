import { ChakraProvider, extendTheme } from "@chakra-ui/react";
import { render as testingLibraryRender } from "@testing-library/react";
import React from "react";
import { BasePlatformTheme } from "./Theme/theme";
var DefaultTheme = extendTheme(BasePlatformTheme);
export var render = function (ui, theme) {
    if (theme === void 0) { theme = DefaultTheme; }
    var wrapper = function (_a) {
        var children = _a.children;
        return (React.createElement(ChakraProvider, { theme: theme }, children));
    };
    return testingLibraryRender(ui, { wrapper: wrapper });
};
