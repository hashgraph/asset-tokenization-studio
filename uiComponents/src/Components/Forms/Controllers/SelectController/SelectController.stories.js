var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import React from "react";
import { SelectController } from "./SelectController";
import { useForm } from "react-hook-form";
var meta = {
    title: "Design System/Forms/Controllers/SelectController",
    component: SelectController,
    args: {
        label: "Hello",
        placeholder: "Hello",
        id: "Name"
    },
    argTypes: {},
    parameters: {
        docs: {}
    }
};
export default meta;
var Template = function (args) {
    var form = useForm({ mode: "onChange" });
    return (React.createElement(SelectController, __assign({}, args, { id: "field", label: "Label example", control: form.control, options: [
            { value: 1, label: "One" },
            { value: 2, label: "Two" },
            { value: 3, label: "Three" },
        ], onChange: function (val) {
            console.log("onChange fired", val);
        } })));
};
export var NoValidations = Template.bind({});
NoValidations.args = {};
export var WithValidations = Template.bind({});
WithValidations.args = {
    rules: {
        required: true,
        validate: { notSecond: function (val) { return val !== 2 || "Option not valid"; } }
    }
};
export var HideErrors = Template.bind({});
HideErrors.args = {
    rules: {
        required: true,
        validate: { notSecond: function (val) { return val !== 2 || "Option not valid"; } }
    },
    showErrors: false
};
export var SetsFullOption = Template.bind({});
SetsFullOption.args = {
    setsFullOption: true
};
