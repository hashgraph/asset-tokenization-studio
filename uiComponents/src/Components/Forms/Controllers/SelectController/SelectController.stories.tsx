import type { Meta, StoryFn } from "@storybook/react";
import React from "react";

import { SelectController } from "./SelectController";
import { useForm } from "react-hook-form";

const meta = {
  title: "Design System/Forms/Controllers/SelectController",
  component: SelectController,
  args: {
    label: "Hello",
    placeholder: "Hello",
    id: "Name",
  },
  argTypes: {},
  parameters: {
    docs: {},
  },
} as Meta<typeof SelectController>;
export default meta;

interface FieldValues {
  field: number;
}

const Template: StoryFn<typeof SelectController> = (args) => {
  const form = useForm<FieldValues>({ mode: "onChange" });
  return (
    <SelectController
      {...args}
      id="field"
      label="Label example"
      control={form.control}
      options={[
        { value: 1, label: "One" },
        { value: 2, label: "Two" },
        { value: 3, label: "Three" },
      ]}
      onChange={(val) => {
        console.log("onChange fired", val);
      }}
    />
  );
};

export const NoValidations = Template.bind({});
NoValidations.args = {};

export const WithValidations = Template.bind({});
WithValidations.args = {
  rules: {
    required: true,
    validate: { notSecond: (val: number) => val !== 2 || "Option not valid" },
  },
};

export const HideErrors = Template.bind({});
HideErrors.args = {
  rules: {
    required: true,
    validate: { notSecond: (val: number) => val !== 2 || "Option not valid" },
  },
  showErrors: false,
};

export const SetsFullOption = Template.bind({});
SetsFullOption.args = {
  setsFullOption: true,
};
