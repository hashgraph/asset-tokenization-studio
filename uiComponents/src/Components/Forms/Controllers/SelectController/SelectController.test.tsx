import type { FieldValues } from "react-hook-form";
import { useForm } from "react-hook-form";
import type { SelectControllerProps } from "./SelectController";
import { SelectController as Select } from "./SelectController";
import { select } from "react-select-event";
import { render } from "@/test-utils";
import React from "react";
import { waitFor } from "@testing-library/react";
const defaultProps = {
  id: "test",
  options: [
    { value: "value1", label: "label1" },
    { value: "value2", label: "label2" },
  ],
  label: "prueba",
  placeholder: "testPlaceholder",
  showErrors: true,
};

//@ts-ignore
let form;
const SelectController = (
  props: Omit<SelectControllerProps<FieldValues>, "control">
) => {
  form = useForm();

  return (
    <form data-testid="form">
      <Select control={form.control} {...props} />
    </form>
  );
};

describe(`<SelectController />`, () => {
  const factoryComponent = (
    props: Omit<SelectControllerProps<FieldValues>, "control"> = defaultProps
  ) => render(<SelectController {...props} />);

  test("render correctly", () => {
    const component = factoryComponent();

    expect(component).toMatchSnapshot();
  });

  test("validate changes in form", async () => {
    const component = factoryComponent();

    await waitFor(async () => {
      await select(component.container.querySelector("input#test")!, "label2");
    });
  });
});
