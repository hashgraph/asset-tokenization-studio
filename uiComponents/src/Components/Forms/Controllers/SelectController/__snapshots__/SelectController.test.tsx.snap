// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectController /> render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    class="chakra-ui-light"
  >
    <div>
      <form
        data-testid="form"
      >
        <div
          class="chakra-form-control css-1kxonj9"
          role="group"
        >
          <label
            class="chakra-form__label css-31pxj9"
            for="test"
            id="field-:r0:-label"
          >
            <span
              class="css-1e9661m"
              id="label"
            >
              prueba
            </span>
            <div
              class=" css-b62m3t-container"
              id="test"
            >
              <span
                class="css-1f43avz-a11yText-A11yText"
                id="react-select-2-live-region"
              />
              <span
                aria-atomic="false"
                aria-live="polite"
                aria-relevant="additions text"
                class="css-1f43avz-a11yText-A11yText"
              />
              <div
                class=" css-13cymwt-control"
              >
                <div
                  class=" css-1fdsijx-ValueContainer"
                >
                  <div
                    class=" css-1jqq78o-placeholder"
                    id="react-select-2-placeholder"
                  >
                    testPlaceholder
                  </div>
                  <div
                    class=" css-qbdosj-Input"
                    data-value=""
                  >
                    <input
                      aria-autocomplete="list"
                      aria-describedby="react-select-2-placeholder"
                      aria-expanded="false"
                      aria-haspopup="true"
                      autocapitalize="none"
                      autocomplete="off"
                      autocorrect="off"
                      class=""
                      id="test"
                      role="combobox"
                      spellcheck="false"
                      style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                      tabindex="0"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
                <div
                  class=" css-1hb7zxy-IndicatorsContainer"
                >
                  <span
                    class=" css-1u9des2-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    class=" css-1xc3v61-indicatorContainer"
                  >
                    <svg
                      aria-hidden="true"
                      class="css-tj5bde-Svg"
                      focusable="false"
                      height="20"
                      viewBox="0 0 20 20"
                      width="20"
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
              <input
                name="test"
                type="hidden"
                value=""
              />
            </div>
          </label>
        </div>
      </form>
      <span
        hidden=""
        id="__chakra_env"
      />
    </div>
    <div
      class="chakra-portal"
    >
      <div
        aria-live="polite"
        id="chakra-toast-manager-top"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <div
        aria-live="polite"
        id="chakra-toast-manager-top-left"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-live="polite"
        id="chakra-toast-manager-top-right"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-live="polite"
        id="chakra-toast-manager-bottom-left"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-live="polite"
        id="chakra-toast-manager-bottom"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <div
        aria-live="polite"
        id="chakra-toast-manager-bottom-right"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
    </div>
  </body>,
  "container": <div>
    <form
      data-testid="form"
    >
      <div
        class="chakra-form-control css-1kxonj9"
        role="group"
      >
        <label
          class="chakra-form__label css-31pxj9"
          for="test"
          id="field-:r0:-label"
        >
          <span
            class="css-1e9661m"
            id="label"
          >
            prueba
          </span>
          <div
            class=" css-b62m3t-container"
            id="test"
          >
            <span
              class="css-1f43avz-a11yText-A11yText"
              id="react-select-2-live-region"
            />
            <span
              aria-atomic="false"
              aria-live="polite"
              aria-relevant="additions text"
              class="css-1f43avz-a11yText-A11yText"
            />
            <div
              class=" css-13cymwt-control"
            >
              <div
                class=" css-1fdsijx-ValueContainer"
              >
                <div
                  class=" css-1jqq78o-placeholder"
                  id="react-select-2-placeholder"
                >
                  testPlaceholder
                </div>
                <div
                  class=" css-qbdosj-Input"
                  data-value=""
                >
                  <input
                    aria-autocomplete="list"
                    aria-describedby="react-select-2-placeholder"
                    aria-expanded="false"
                    aria-haspopup="true"
                    autocapitalize="none"
                    autocomplete="off"
                    autocorrect="off"
                    class=""
                    id="test"
                    role="combobox"
                    spellcheck="false"
                    style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                    tabindex="0"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                class=" css-1hb7zxy-IndicatorsContainer"
              >
                <span
                  class=" css-1u9des2-indicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  class=" css-1xc3v61-indicatorContainer"
                >
                  <svg
                    aria-hidden="true"
                    class="css-tj5bde-Svg"
                    focusable="false"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <input
              name="test"
              type="hidden"
              value=""
            />
          </div>
        </label>
      </div>
    </form>
    <span
      hidden=""
      id="__chakra_env"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
